field/TinyLang/Field/UParser.hs

```diff
@@ -0,0 +1,258 @@
+@
+
+== Operator Precedence
+  Precedence: "not" > "xor" > "and" > "or"  (but use parentheses anyway).
+  if-then-else has to be parenthesised unless it's at the very top.
+
+  Precedence for numeric operators is standard:  {neg,inv} > {*,/} > {+,- }.
+  Things like "neg inv 5" are illegal: use parentheses.
+
+  The code is based on the tutorial at
+  https://markkarpov.com/megaparsec/parsing-simple-imperative-language.html
+
+  See also https://markkarpov.com/megaparsec/megaparsec.html
+-}
+
+
+module TinyLang.Field.UParser where
+
+import           Control.Applicative.Combinators hiding (many, sepBy, some)
+import           Text.Megaparsec
+import           Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer      as L
+import           Data.Void
+import qualified Data.Set                        as S
+import           GHC.Unicode ( isLower
+                             , isDigit)
+
+type Parser = Parsec Void String
+
+type Identifier = String
+
+data Const
+  = CBool  Bool
+  | CInt   Integer
+  | CVec   [Bool]
+
+data Var = Var Identifier
+
+data Expr
+  = EConst     Const
+  | EVar       Var
+  | EAppBinOp  UnOp      Expr
+  | EAppUnOp   BinOp     Expr
+  | EStatement Statement Expr
+  | EIf        Expr      Expr Expr
+
+data BinOp
+  = Or
+  | And
+  | Xor
+  | FEq
+  | FLe
+  | FGe
+  | FGt
+  | Add
+  | Sub
+  | Mul
+  | Div
+  | BAt
+
+data UnOp
+  = Not
+  | Neq0
+  | Neg
+  | Inv
+  | Unp
+  
+data Statement
+  = ELet    Var Expr
+  | EAssert Expr
+
+-- Lexer
+sc :: Parser ()
+sc = L.space space1 empty empty
+
+lexeme :: Parser a -> Parser a
+lexeme = L.lexeme sc
+
+symbol :: String -> Parser String
+symbol = L.symbol sc
+
+
+-- Identifier Character
+isIdentifierChar :: Char -> Bool
+isIdentifierChar c =
+  or $ map (\f -> f c) [isLower , isDigit, (=='_'), (=='\'')]
+
+
+identifierChar :: (MonadParsec e s m, Token s ~ Char) => m (Token s)
+identifierChar = satisfy isIdentifierChar <?> "identifier character [a-z0-9_']"
+{-# INLINE identifierChar #-}
+
+pKeyword :: String -> Parser String
+pKeyword keyword = lexeme (string keyword <* notFollowedBy identifierChar)
+
+keywords :: S.Set String
+keywords =
+  S.fromList 
+    [ "T", "F"
+    , "not", "and", "or", "xor"
+    , "neq0", "neg", "inv"
+    , "let"
+    , "if", "then", "else"
+    , "for", "do", "end"
+    , "unpack"
+    ]
+
+isKeyword :: String -> Bool
+isKeyword = (`S.member` keywords)
+
+pIdentifier :: Parser String
+pIdentifier =
+  lexeme $ do
+  prefix     <- option "" (string "?" <|> string "#")
+  identifier <- (:) <$> lowerChar <*> many identifierChar
+  pure $ prefix ++ identifier
+
+pBoolLiteral :: Parser String
+pBoolLiteral =
+  lexeme (fmap pure (satisfy (\x -> x == 'T' || x == 'F') <?> "T or F"))
+
+pIntLiteral :: Parser String
+pIntLiteral =
+  lexeme (some digitChar)
+
+pVecLiteral :: Parser String
+pVecLiteral =
+  lexeme $
+  between
+  (symbol "{")
+  (symbol "}")
+  ((pBoolLiteral `sepBy` (symbol ",")) >>= (pure . concat))
```

Comment:
``concat <$> (pBoolLiteral `sepBy` symbol ",")``?

Reply:
Yep, I am still getting used to applicative-style programming.